src/main/kotlin/org/jitsi/nlj/rtp/codec/vp9/Vp9Packet.kt

```diff
@@ -0,0 +1,305 @@
+        isEndOfFrame = null,
+        encodingIndex = null,
+        pictureId = null,
+        TL0PICIDX = null
+    )
+
+    override val isKeyframe: Boolean =
+        isKeyframe ?: DePacketizer.VP9PayloadDescriptor.isKeyFrame(this.buffer, payloadOffset, payloadLength)
+
+    override val isStartOfFrame: Boolean =
+        isStartOfFrame ?: DePacketizer.VP9PayloadDescriptor.isStartOfFrame(buffer, payloadOffset, payloadLength)
+
+    override val isEndOfFrame: Boolean =
+        isEndOfFrame ?: DePacketizer.VP9PayloadDescriptor.isEndOfFrame(buffer, payloadOffset, payloadLength)
+
+    override val layerId: Int
+        get() = if (hasLayerIndices) RtpLayerDesc.getIndex(0, spatialLayerIndex, temporalLayerIndex) else super.layerId
+
+    /** End of VP9 picture is the marker bit. Note frame/picture distinction. */
+    /* TODO: not sure this should be the override from [ParsedVideoPacket] */
+    val isEndOfPicture: Boolean
+        /** This uses [get] rather than initialization because [isMarked] is a var. */
+        get() = isMarked
+
+    val hasLayerIndices =
+        DePacketizer.VP9PayloadDescriptor.hasLayerIndices(buffer, payloadOffset, payloadLength)
+
+    val hasPictureId =
+        DePacketizer.VP9PayloadDescriptor.hasPictureId(buffer, payloadOffset, payloadLength)
+
+    val hasExtendedPictureId =
+        DePacketizer.VP9PayloadDescriptor.hasExtendedPictureId(buffer, payloadOffset, payloadLength)
+
+    val hasTL0PICIDX =
+        DePacketizer.VP9PayloadDescriptor.hasTL0PICIDX(buffer, payloadOffset, payloadLength)
+
+    val hasScalabilityStructure =
+        DePacketizer.VP9PayloadDescriptor.hasScalabilityStructure(buffer, payloadOffset, payloadLength)
+
+    val isUpperLevelReference =
+        DePacketizer.VP9PayloadDescriptor.isUpperLevelReference(buffer, payloadOffset, payloadLength)
+
+    val isInterPicturePredicted =
+        DePacketizer.VP9PayloadDescriptor.isInterPicturePredicted(buffer, payloadOffset, payloadLength)
+
+    private var _TL0PICIDX =
+        TL0PICIDX ?: DePacketizer.VP9PayloadDescriptor.getTL0PICIDX(buffer, payloadOffset, payloadLength)
+
+    var TL0PICIDX: Int
+        get() = _TL0PICIDX
+        set(newValue) {
+            if (newValue != -1 && !DePacketizer.VP9PayloadDescriptor.setTL0PICIDX(
+                    buffer, payloadOffset, payloadLength, newValue)) {
+                logger.cwarn { "Failed to set the TL0PICIDX of a VP9 packet." }
+            }
+            _TL0PICIDX = newValue
+        }
+
+    private var _pictureId =
+        pictureId ?: DePacketizer.VP9PayloadDescriptor.getPictureId(buffer, payloadOffset, payloadLength)
+
+    var pictureId: Int
+        get() = _pictureId
+        set(newValue) {
+            if (!DePacketizer.VP9PayloadDescriptor.setExtendedPictureId(
+                    buffer, payloadOffset, payloadLength, newValue)) {
+                logger.cwarn { "Failed to set the picture id of a VP9 packet." }
+            }
+            _pictureId = newValue
+        }
```

Comment:
I think all this code around `pictureId` is the same as:

```
val pictureId: Int by observable(pictureId ?: DePacketizer.VP9PayloadDescriptor.getPictureId(buffer, payloadOffset, payloadLength)) { _, _, newValue -> {
    if (!DePacketizer.VP9PayloadDescriptor.setExtendedPictureId(buffer, payloadOffset, payloadLength, newValue)) {
                logger.cwarn { "Failed to set the picture id of a VP9 packet." }
            }
}
```
And should maybe even use `vetoable` to refuse the change if it couldn't be updated?

Reply:
It is, yes, but with the Vp8Packet what we determined is that per-packet observables of primitive types on actually imposed quite a lot of overhead.  This is the VP9 translation of https://github.com/jitsi/jitsi-media-transform/pull/259 .

Reply:
Cool, glad you remembered that :)